pkg/webui/console/views/device-add/wizard/index.js

```diff
@@ -0,0 +1,193 @@
+// limitations under the License.
+
+import React from 'react'
+
+import Wizard from '@ttn-lw/components/wizard'
+
+import sharedMessages from '@ttn-lw/lib/shared-messages'
+import getHostnameFromUrl from '@ttn-lw/lib/host-from-url'
+import PropTypes from '@ttn-lw/lib/prop-types'
+
+import { ACTIVATION_MODES } from '@console/lib/device-utils'
+
+import m from '../messages'
+import {
+  getActivationMode,
+  getLorawanVersion,
+  getApplicationServerAddress,
+  getNetworkServerAddress,
+  getJoinServerAddress,
+} from '../utils'
+
+import BasicSettingsForm from './basic-settings-form'
+import NetworkSettingsForm from './network-settings-form'
+import ApplicationSettingsForm from './application-settings-form'
+import JoinSettingsForm from './join-settings-form'
+import Prompt from './prompt'
+
+const DeviceWizard = React.memo(props => {
+  const {
+    configuration,
+    asConfig,
+    nsConfig,
+    jsConfig,
+    mayEditKeys,
+    match,
+    createDevice,
+    createDeviceSuccess,
+    rollbackProgress,
+  } = props
+
+  const [completed, setCompleted] = React.useState(false)
+  const [error, setError] = React.useState('')
+
+  const handleWizardComplete = React.useCallback(
+    async values => {
+      try {
+        await createDevice(values)
+        setCompleted(true)
+        return createDeviceSuccess(values)
+      } catch (error) {
+        setError(error)
+      }
+    },
+    [createDevice, createDeviceSuccess],
+  )
+
+  const handleBlockNavigation = React.useCallback(
+    location => !location.pathname.endsWith('/steps'),
+    [],
+  )
+
+  const handlePromptApprove = React.useCallback(
+    location => {
+      rollbackProgress(location.pathname)
+    },
+    [rollbackProgress],
+  )
+
+  return React.createElement(
+    Wizard,
+    {
+      initialValues: configuration,
+      onComplete: handleWizardComplete,
+      completeMessage: sharedMessages.addDevice,
+    },
+    ({ snapshot, dirty }) => {
+      const stepperStatus = Boolean(error) ? 'failure' : 'current'
+      const activationMode = getActivationMode(snapshot)
+      const lorawanVersion = getLorawanVersion(snapshot)
+
+      const showNetworkStep =
+        nsConfig.enabled &&
+        getNetworkServerAddress(snapshot) === getHostnameFromUrl(nsConfig.base_url) &&
+        activationMode !== ACTIVATION_MODES.NONE
+      const showApplicationStep =
+        asConfig.enabled &&
+        getApplicationServerAddress(snapshot) === getHostnameFromUrl(asConfig.base_url) &&
+        (activationMode === ACTIVATION_MODES.ABP || activationMode === ACTIVATION_MODES.MULTICAST)
+      const showJoinStep =
+        getJoinServerAddress(snapshot) === getHostnameFromUrl(jsConfig.base_url) &&
+        activationMode === ACTIVATION_MODES.OTAA
+
+      return (
+        <>
+          <Prompt
+            when={dirty && !completed}
+            shouldBlockNavigation={handleBlockNavigation}
+            onApprove={handlePromptApprove}
+          />
```

Comment:
This prompting could really be a default feature of the wizard component. I can't think of any scenario where we don't want to notify users of unsaved changes.

It would also take away some of the complexity here.